shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/presentation/StudyPlanWidgetReducer.kt

```diff
@@ -114,61 +101,39 @@ class StudyPlanWidgetReducer : StateReducer<State, Message, Action> {
                 )
         } ?: (state to emptySet())
 
-    private fun coldContentFetch(state: State, message: Message.Initialize): StudyPlanWidgetReducerResult =
+    private fun coldContentFetch(state: State, message: InternalMessage.Initialize): StudyPlanWidgetReducerResult =
         if (state.sectionsStatus == StudyPlanWidgetFeature.ContentStatus.IDLE ||
             state.sectionsStatus == StudyPlanWidgetFeature.ContentStatus.ERROR && message.forceUpdate
         ) {
             State(sectionsStatus = StudyPlanWidgetFeature.ContentStatus.LOADING) to
-                setOf(InternalAction.FetchStudyPlan(), InternalAction.FetchProfile)
+                getContentFetchActions(forceUpdate = message.forceUpdate)
         } else {
             state to emptySet()
         }
 
-    private fun handleStudyPlanFetchSuccess(
-        state: State,
-        message: StudyPlanWidgetFeature.StudyPlanFetchResult.Success
-    ): StudyPlanWidgetReducerResult {
-        val actions = if (message.studyPlan.status != StudyPlanStatus.READY) {
-            setOf(
-                InternalAction.FetchStudyPlan(
-                    delayBeforeFetching = STUDY_PLAN_FETCH_INTERVAL * message.attemptNumber,
-                    attemptNumber = message.attemptNumber + 1
-                )
-            )
-        } else {
-            setOfNotNull(
-                InternalAction.FetchSections(message.studyPlan.sections),
-                message.studyPlan.trackId?.let(InternalAction::FetchTrack)
-            )
-        }
-
-        return if (message.showLoadingIndicators) {
-            state.copy(
-                studyPlan = message.studyPlan,
-                sectionsStatus = StudyPlanWidgetFeature.ContentStatus.LOADING
-            ) to actions
-        } else {
-            state.copy(studyPlan = message.studyPlan) to actions
-        }
-    }
+    private fun getContentFetchActions(forceUpdate: Boolean): Set<Action> =
+        setOf(
+            InternalAction.FetchLearningActivitiesWithSections(),
+            InternalAction.FetchProfile,
+            InternalAction.UpdateCurrentStudyPlanState(forceUpdate)
+        )
 
-    private fun handleSectionsFetchSuccess(
+    private fun handleLearningActivitiesWithSectionsFetchSuccess(
         state: State,
-        message: StudyPlanWidgetFeature.SectionsFetchResult.Success
+        message: InternalMessage.FetchLearningActivitiesWithSectionsSuccess
     ): StudyPlanWidgetReducerResult {
-        if (state.studyPlan == null) {
-            return state to emptySet()
-        }
-
-        val sortedSupportedSections = message.sections
-            .filter { it.isVisible && StudyPlanSectionType.supportedTypes().contains(it.type) }
-            .sortedBy { state.studyPlan.sections.indexOf(it.id) }
+        val visibleSections = message.studyPlanSections.filter { it.isVisible }
+        val currentSectionId = visibleSections.firstOrNull()?.id
```

Comment:
We should determine current section in other way. Current section is the section that contains learning activities that was returned by that API. In current realization there could be situation when API returned activities from not first visible section. In this way we should hide all visible section that above of current section. For example, the first visible section contains only not supported activities.

Reply:
OMG, why like that 🤯

Reply:
I really wanted to do it on backend (pass `is_visible = false` for sections above), but it wasn't easy in realization, sorry(